atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraRequestExceptionHandler.java

```diff
@@ -53,109 +54,162 @@
-            log.info("Retrying with fast failover a query intended for host {}.",
-                    SafeArg.of("hostName", CassandraLogHelper.host(hostTried)));
-            req.giveUpOnPreferredHost();
+        if (numberOfAttempts >= maxTriesTotal.get()) {
+            throw (K) logAndReturnException(numberOfAttempts, ex);
         }
+
+        if (shouldBlacklist(ex, numberOfAttempts)) {
+            blacklist.add(hostTried);
+        }
+
+        logNumberOfAttempts(ex, numberOfAttempts);
+        handleBackoff(req, hostTried, ex);
+        handleRetryOnDifferentHosts(req, hostTried, ex);
     }
 
     @SuppressWarnings("unchecked")
-    private <K extends Exception> void handleExceptionInternal(
-            RetryableCassandraRequest<?, K> req, InetSocketAddress host, Exception ex) throws K {
-        if (isRetriableException(ex) || isRetriableWithBackoffException(ex) || isFastFailoverException(ex)) {
-            if (req.getNumberOfAttempts() >= maxTriesTotal.get()) {
-                if (ex instanceof TTransportException
-                        && ex.getCause() != null
-                        && (ex.getCause().getClass() == SocketException.class)) {
-                    log.error(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts(), ex);
-                    String errorMsg =
-                            MessageFormatter.format(CONNECTION_FAILURE_MSG, req.getNumberOfAttempts()).getMessage();
-                    throw (K) new TTransportException(((TTransportException) ex).getType(), errorMsg, ex);
-                } else {
-                    log.error("Tried to connect to cassandra {} times.",
-                            SafeArg.of("numTries", req.getNumberOfAttempts()), ex);
-                    throw (K) ex;
-                }
-            } else {
-                // Only log the actual exception the first time
-                if (req.getNumberOfAttempts() > 1) {
-                    log.info("Error occurred talking to cassandra. Attempt {} of {}. Exception message was: {} : {}",
-                            SafeArg.of("numTries", req.getNumberOfAttempts()),
-                            SafeArg.of("maxTotalTries", maxTriesTotal),
-                            SafeArg.of("exceptionClass", ex.getClass().getTypeName()),
-                            UnsafeArg.of("exceptionMessage", ex.getMessage()));
-                } else {
-                    log.info("Error occurred talking to cassandra. Attempt {} of {}.",
-                            SafeArg.of("numTries", req.getNumberOfAttempts()),
-                            SafeArg.of("maxTotalTries", maxTriesTotal),
-                            ex);
-                }
-                if (isConnectionException(ex) && req.getNumberOfAttempts() >= maxTriesSameHost.get()) {
-                    blacklist.add(host);
-                }
-            }
+    private <K extends Exception> K logAndReturnException(int numberOfAttempts, Exception ex) {
```

Comment:
Can we change this to logAndThrowException and then actually throw the Exception.  This way callers don't have to remember to do the throwing (but the code is clearer if they do throw the result still).